### PR TITLE
Feature/remove cowboy

### PR DESCRIPTION
**Pull request by scohen**
Removes cowboy as a dependency, replacing it with a simple HTTP server using ranch.

### CODE REVIEW

**Comment by scohen** [comment]
Also,  since I'm making breaking changes (we don't send the request to the init functions), i'm wondering if I should make a `loqui_push` callback.

File: ex/loqui/lib/loqui/server.ex [hunk 1]
```diff
@@ -0,0 +1,136 @@
+defmodule Loqui.Server do
+
+  defmodule Http do
+    defmodule State do
+      defstruct [:handler, :path, :transport, :socket, :buffer, :handler_opts]
+
+      def new(socket, transport, opts) do
+        %__MODULE__{
+          path: Keyword.fetch!(opts, :loqui_path),
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/server.ex]: If these are required, should they be arguments instead of options?

File: ex/loqui/lib/loqui/server.ex [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    end
+
+    defp handle_tcp_data(extra_data, %{socket: sock, path: loqui_path, transport: transport, buffer: buffer} = state) do
+      with {:ok, ^loqui_path, {headers, _}} <- try_parse_request(buffer <> extra_data),
+           "loqui" <- :proplists.get_value("upgrade", headers) do
+
+        response = :cow_http.response(101, :"HTTP/1.1", [{"upgrade", "loqui"}])
+        transport.send(sock, response)
+        RanchProtocol.upgrade(sock, transport, state.handler, state.handler_opts)
+        :ok
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/server.ex]: Should this function return the result of `RanchProtocol.upgrade`? I think we're going to swallow some errors if we don't.

File: ex/loqui/lib/loqui/server.ex [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+          transport.send(sock, response)
+          :ok
+      end
+    end
+
+    defp try_parse_request(data) when byte_size(data) >= @max_payload,
+      do: {:error, :payload_too_large}
+
+    defp try_parse_request(data) do
+      with true <- String.ends_with?(data, "\r\n\r\n"),
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/server.ex]: We are guaranteed that the client won't send a message after the initial upgrade request?

File: ex/loqui/lib/loqui/ranch_protocol.ex [hunk 4]
```diff
@@ -313,22 +299,26 @@ defmodule Loqui.CowboyProtocol do
   end
 
   @spec handler_push(state, binary) :: :ok
+  defp handler_push(%{handler: handler, codec: codec, handler_supports_push: true}=state, request) do
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/ranch_protocol.ex]: Was this just not used in the past? 

If the handler doesn't support push and the client tries to do a push, what will the client receive?

Comment (scohen) [reply ex/loqui/lib/loqui/ranch_protocol.ex]: The prior implementation (which is still there for handlers that don't have this function) is to call `loqui_request`.

File: ex/loqui/lib/loqui/server.ex (hunk 1 above)
Comment (scohen) [reply ex/loqui/lib/loqui/server.ex]: This is internal to the HTTP module. The external facing `Server.start_link` function has `loqui_path` as an argument.

File: ex/loqui/lib/loqui/ranch_protocol.ex (hunk 4 above)
Comment (jhowarth) [reply ex/loqui/lib/loqui/ranch_protocol.ex]: I see, I misread 👍

File: ex/loqui/lib/loqui/server.ex (hunk 2 above)
Comment (scohen) [reply ex/loqui/lib/loqui/server.ex]: ok

File: ex/loqui/lib/loqui/server.ex (hunk 3 above)
Comment (scohen) [reply ex/loqui/lib/loqui/server.ex]: Aren't we in the same boat with Cowboy? Am I missing some sort of implicit downgrade that was happening? How can the protocol tell the difference betwen loqui sending a string whose contents are a valid HTTP request and actually sending a HTTP request? 

I guess my point here is that if the client suddenly starts speaking HTTP again, it'll be disconnected, which seems like correct behavior in this case.

File: ex/loqui/lib/loqui/ranch_protocol.ex [hunk 5]
```diff
@@ -396,9 +380,9 @@ defmodule Loqui.CowboyProtocol do
     end
   end
 
-  @spec address(req) :: String.t
-  defp address(req) do
-    {{address, _}, _} = :cowboy_req.peer(req)
+  @spec address(map) :: String.t
+  defp address(state) do
+    {:ok, {address, _port}} = state.transport.peername(state.socket_pid)
```
[the remark this answers is not in the document]
Comment (scohen) [reply ex/loqui/lib/loqui/ranch_protocol.ex]: IDK, I can change.

File: ex/loqui/lib/loqui/server.ex (hunk 3 above)
Comment (scohen) [reply ex/loqui/lib/loqui/server.ex]: Or, are you saying that the message will be attached to the initial HTTP request?

File: ex/loqui/lib/loqui/ranch_protocol.ex (hunk 4 above)
Comment (scohen) [reply ex/loqui/lib/loqui/ranch_protocol.ex]: It wasn't shown in the diff, it's easy to not see.

File: ex/loqui/lib/loqui/server.ex (hunk 1 above)
Comment (scohen) [reply ex/loqui/lib/loqui/server.ex]: I think this is due to how ranch starts a listener. you get `start_link(ref, socket, transport, opts)`, and I have to pack everything in the opts.

File: ex/loqui/lib/loqui/server.ex [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+          Logger.info("Client #{ip_address(sock)} closed.")
+          exit(:normal)
+      end
+    end
+
+    defp handle_tcp_data(extra_data, %{socket: sock, path: loqui_path, transport: transport, buffer: buffer} = state) do
+      with {:ok, ^loqui_path, {headers, _}} <- try_parse_request(buffer <> extra_data),
+           "loqui" <- :proplists.get_value("upgrade", headers) do
+
+        response = :cow_http.response(101, :"HTTP/1.1", [{"upgrade", "loqui"}])
```
[the remark this answers is not in the document]
Comment (scohen) [reply ex/loqui/lib/loqui/server.ex]: IDK, I found those functions helpful, especially the header parsing stuff.

File: ex/loqui/lib/loqui/server.ex (hunk 3 above)
Comment (jhowarth) [reply ex/loqui/lib/loqui/server.ex]: I'm saying another message will be attached to the initial HTTP request. Then we will keep accepting data. It's an edge case, not sure if it's worth addressing.

Comment (scohen) [reply ex/loqui/lib/loqui/server.ex]: That violates the protocol, no? They should respond with a loqui negotiation, not another HTTP request.

**Pull request merged**